server/worker/jira.go

```diff
@@ -42,73 +50,256 @@ View the affected software and more affected hosts:
+		}
+		for _, intg := range ac.Integrations.Jira {
+			if (intgType == intgTypeVuln && intg.EnableSoftwareVulnerabilities) ||
+				(intgType == intgTypeFailingPolicy && intg.EnableFailingPolicies) {
+				opts = &externalsvc.JiraOptions{
+					BaseURL:           intg.URL,
+					BasicAuthUsername: intg.Username,
+					BasicAuthPassword: intg.APIToken,
+					ProjectKey:        intg.ProjectKey,
+				}
+				break
+			}
+		}
+	}
+
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	if j.clientsCache == nil {
+		j.clientsCache = make(map[string]JiraClient)
+	}
+	if opts == nil {
+		// no integration configured, clear any existing one
+		delete(j.clientsCache, key)
+		return nil, nil
+	}
+
+	// check if the existing one can be reused
+	if cli := j.clientsCache[key]; cli != nil && cli.JiraConfigMatches(opts) {
+		return cli, nil
+	}
+
+	// otherwise create a new one
+	cli, err := j.NewClientFunc(opts)
+	if err != nil {
+		return nil, err
+	}
+	j.clientsCache[key] = cli
+	return cli, nil
+}
+
+// jiraArgs are the arguments for the Jira integration job.
+type jiraArgs struct {
+	CVE           string             `json:"cve,omitempty"`
+	FailingPolicy *failingPolicyArgs `json:"failing_policy,omitempty"`
+}
+
+func (a *jiraArgs) integrationType() string {
+	if a.FailingPolicy == nil {
+		return intgTypeVuln
+	}
+	return intgTypeFailingPolicy
 }
 
 // Run executes the jira job.
 func (j *Jira) Run(ctx context.Context, argsJSON json.RawMessage) error {
-	var args JiraArgs
+	var args jiraArgs
 	if err := json.Unmarshal(argsJSON, &args); err != nil {
 		return ctxerr.Wrap(ctx, err, "unmarshal args")
 	}
 
+	cli, err := j.getClient(ctx, args)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "get Jira client")
+	}
+	if cli == nil {
+		// this message was queued when an integration was enabled, but since
+		// then it has been disabled, so return success to mark the message
+		// as processed.
+		return nil
+	}
+
+	switch intgType := args.integrationType(); intgType {
+	case intgTypeVuln:
+		return j.runVuln(ctx, cli, args)
+	case intgTypeFailingPolicy:
+		return j.runFailingPolicies(ctx, cli, args)
+	default:
+		return ctxerr.Errorf(ctx, "unknown integration type: %v", intgType)
+	}
+}
+
+func (j *Jira) runVuln(ctx context.Context, cli JiraClient, args jiraArgs) error {
 	hosts, err := j.Datastore.HostsByCVE(ctx, args.CVE)
 	if err != nil {
 		return ctxerr.Wrap(ctx, err, "find hosts by cve")
 	}
 
-	tmplArgs := jiraTemplateArgs{
+	tplArgs := &jiraVulnTplArgs{
 		NVDURL:   nvdCVEURL,
 		FleetURL: j.FleetURL,
 		CVE:      args.CVE,
 		Hosts:    hosts,
 	}
 
+	createdIssue, err := j.createTemplatedIssue(ctx, cli, jiraTemplates.VulnSummary, jiraTemplates.VulnDescription, tplArgs)
+	if err != nil {
+		return err
+	}
+	level.Debug(j.Log).Log(
+		"msg", "created jira issue for cve",
+		"cve", args.CVE,
+		"issue_id", createdIssue.ID,
+		"issue_key", createdIssue.Key,
+	)
+	return nil
+}
+
+func (j *Jira) runFailingPolicies(ctx context.Context, cli JiraClient, args jiraArgs) error {
```

Comment:
Am I correct that this is intended to be called one policy id at a time?  If so, what do you think of renaming it to `runFailingPolicyJob` or something like that?